code/source/Client/ClientAuthApi.cpp

```diff
@@ -0,0 +1,731 @@
+#include "stdafx.h"
+#include "ClientAuthApi.h"
+#include "AuthContext.h"
+
+namespace PlayFab
+{
+
+using namespace ClientModels;
+
+ClientAuthAPI::ClientAuthAPI() :
+    m_settings{ MakeShared<PlayFabApiSettings>() },
+    m_httpClient{ m_settings }
+{
+}
+
+void ClientAuthAPI::LoginWithAndroidDeviceID(
+    const PlayFabClientLoginWithAndroidDeviceIDRequest& request,
+    const TaskQueue& queue,
+    const ProcessApiCallback<SharedPtr<AuthContext>> callback,
+    const ErrorCallback errorCallback
+) const
+{
+    // TODO find a cleaner way to configure the titleId
+    m_settings->titleId = request.titleId;
+    UnorderedMap<String, String> headers;
+
+    // TODO bug: There is a lifetime issue with capturing this here since the client owns the object
+    auto callComplete = [ this, callback, errorCallback ](const HttpResult& httpResult)
+    {
+        if (SUCCEEDED(httpResult.networkErrorCode) && httpResult.serviceResponse.HttpCode == 200)
+        {
+            LoginResult resultModel;
+            resultModel.FromJson(httpResult.serviceResponse.Data);
+            callback(MakeShared<AuthContext>(resultModel));
+        }
+        else if (errorCallback)
+        {
+            errorCallback(httpResult.serviceResponse);
+        }
+    };
+
+    m_httpClient.MakePostRequest(
+        "/Client/LoginWithAndroidDeviceID",
+        headers,
+        JsonUtils::ToJson(request),
+        queue,
+        callComplete
+    );
+}
+void ClientAuthAPI::LoginWithApple(
+    const PlayFabClientLoginWithAppleRequest& request,
+    const TaskQueue& queue,
+    const ProcessApiCallback<SharedPtr<AuthContext>> callback,
+    const ErrorCallback errorCallback
+) const
+{
+    // TODO find a cleaner way to configure the titleId
+    m_settings->titleId = request.titleId;
+    UnorderedMap<String, String> headers;
+
+    // TODO bug: There is a lifetime issue with capturing this here since the client owns the object
+    auto callComplete = [ this, callback, errorCallback ](const HttpResult& httpResult)
+    {
+        if (SUCCEEDED(httpResult.networkErrorCode) && httpResult.serviceResponse.HttpCode == 200)
+        {
+            LoginResult resultModel;
+            resultModel.FromJson(httpResult.serviceResponse.Data);
+            callback(MakeShared<AuthContext>(resultModel));
+        }
+        else if (errorCallback)
+        {
+            errorCallback(httpResult.serviceResponse);
+        }
+    };
+
+    m_httpClient.MakePostRequest(
+        "/Client/LoginWithApple",
+        headers,
+        JsonUtils::ToJson(request),
+        queue,
+        callComplete
+    );
```

Comment:
Can't this live as a helper method that takes in a http url, the request json, and the callbacks?

Reply:
Which part?  Not sure what creating a helper method actually accomplishes here.

---
In reply to: [607292144](https://github.com/jplafonta/XPlatCSdk/pull/2#discussion_r607292144) [](ancestors = 607292144)